### DISCUSSION TITLE
Advice on installing Agate on Ubuntu 22.10 / Intel?

### DISCUSSION THREAD

**Discussion by ygagnon** [mbrubeck/agate#255] [Q&A] [answered]
I just set up a Droplet with Digital Ocean (Ubuntu v22.10 w/ an Intel processor), and am seeking advice on the simplest way to install Agate on it, and to set it up as a service that runs automatically on boot.  

I came across this YouTube video tutorial which seemed fairly comprehensive, but which is labeled as being "OUT OF DATE", which is why I'm inquiring here.

https://www.youtube.com/watch?v=ldPlP4xVIag

Should I be using the "Cargo" method of installation, and if so, are there any specific instructions that I should adhere to given my Ubuntu version / platform?  Or should I be installing from a pre-compiled binary (as demonstrated in the YouTube video), and if so -- which one?  

Lastly, how would I set Agate up as a service that runs automatically on boot?

Thanks,
-- Yvan

**Comment by mbrubeck** [comment]
Either Cargo or the pre-compiled binary is a good way to install the latest version.  Downloading the binary is simpler and faster, unless you already have Rust/Cargo installed for other reasons.

The latest binary release is always available here: https://github.com/mbrubeck/agate/releases

You want the `x86_64-unknown-linux-gnu` binary.

**Reply by ygagnon** [reply]
Ok - thank you.  So are the instructions that are outlined in that YouTube video still mostly the same then?  In other words, can I skip creating the OpenSSL certificates?  And how would I set up Agate to run as a service?

**Reply by mbrubeck** [reply]
Yes, you can skip creating the certificates because Agate will now do this for you, automatically.  The instructions from the video about setting up a systemd service should still work.

**Reply by ygagnon** [reply]
Could you please verify that my agate.service file looks correct?  (Let me know if any corrections or modifications are required, etc.).  Note that I've already created a new "gemini" user, ... created the "bin" and "gemini" directories, ... created my index.gmi file, ... downloaded the binary, ... unzipped it, ... and made it executable. I just purposefully left out my domain (which I've already pointed to my new Digital Ocean droplet):

```
[Unit]
Description=agate
After=network.target

[Service]
User=gemini
Type=simple
ExecStart=/home/gemini/bin/agate.x86_64-unknown-linux-gnu -s --content /home/gemini/gemini/ --addr [::]:1965 --addr 0.0.0.0:1965 --hostname mydomain.com --lang en-US

[Install]
WantedBy=default.target 
```

**Reply by Johann150** [reply]
I do not see anything that is obviously wrong, but of course it depends on a lot of your system's configuration that I don't know. I would assume it should be okay.

You don't strictly need these options, since that is the default behaviour:

```
--addr [::]:1965 --addr 0.0.0.0:1965
```

You may also be interested in the [tools provided for Debian](https://github.com/mbrubeck/agate/tree/master/tools/debian) since they should also be applicable to Ubuntu.

**Reply by ygagnon** [reply]
I'm getting closer to getting this up and running .. but have encountered a new error message that seems to be hinting at a permissions issue when attempting to create a ".certificates" folder:

```
May 04 21:43:05 ubuntu-s-1vcpu-512mb-10gb-sfo3-01 systemd[1]: Started agate.
May 04 21:43:05 ubuntu-s-1vcpu-512mb-10gb-sfo3-01 agate.x86_64-unknown-linux-gnu[48958]: [2023-05-04T21:43:05Z INFO  agate] The certificate directory ".certificates" does not exist, creating it.
May 04 21:43:05 ubuntu-s-1vcpu-512mb-10gb-sfo3-01 agate.x86_64-unknown-linux-gnu[48958]: thread 'main' panicked at 'could not create certificate directory: Os { code: 13, kind: PermissionDenied, message: "Per>
May 04 21:43:05 ubuntu-s-1vcpu-512mb-10gb-sfo3-01 agate.x86_64-unknown-linux-gnu[48958]: note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
May 04 21:43:05 ubuntu-s-1vcpu-512mb-10gb-sfo3-01 systemd[1]: agate.service: Main process exited, code=dumped, status=6/ABRT
May 04 21:43:05 ubuntu-s-1vcpu-512mb-10gb-sfo3-01 systemd[1]: agate.service: Failed with result 'core-dump'.
```

Should I perhaps manually create this .certificates folder?  And if so, where?

**Reply by ygagnon** [reply]
Sorry for the repost.  I'd edited my original one with an update, and suspect that it slipped through the cracks. I'm just really eager to get my Gemini server up and running, and don't really know where else to turn.

-------------------------------------------------------------------------------------------------------------

I'm getting closer to getting this up and running .. but have encountered a new error message that seems to be hinting at a permissions issue when attempting to create a ".certificates" folder:

May 04 21:43:05 ubuntu-s-1vcpu-512mb-10gb-sfo3-01 systemd[1]: Started agate.
May 04 21:43:05 ubuntu-s-1vcpu-512mb-10gb-sfo3-01 agate.x86_64-unknown-linux-gnu[48958]: [2023-05-04T21:43:05Z INFO agate] The certificate directory ".certificates" does not exist, creating it.
May 04 21:43:05 ubuntu-s-1vcpu-512mb-10gb-sfo3-01 agate.x86_64-unknown-linux-gnu[48958]: thread 'main' panicked at 'could not create certificate directory: Os { code: 13, kind: PermissionDenied, message: "Per>
May 04 21:43:05 ubuntu-s-1vcpu-512mb-10gb-sfo3-01 agate.x86_64-unknown-linux-gnu[48958]: note: run with RUST_BACKTRACE=1 environment variable to display a backtrace
May 04 21:43:05 ubuntu-s-1vcpu-512mb-10gb-sfo3-01 systemd[1]: agate.service: Main process exited, code=dumped, status=6/ABRT
May 04 21:43:05 ubuntu-s-1vcpu-512mb-10gb-sfo3-01 systemd[1]: agate.service: Failed with result 'core-dump'.

Should I perhaps manually create this .certificates folder? And if so, where?

**Reply by mbrubeck** [reply, marked as answer]
Make sure agate is trying to create the certificates directory in a writeable location instead of the filesystem root.  You can add `WorkingDirectory=~` to your `[Service]` config so it is not running in the root directory, or you can pass `--certs /home/gemini/certificates` as part of your `ExecStart` command.

**Reply by ygagnon** [reply]
Yesss!!! Creating the new folder, making it writable, and then adding the --certs flag to my ExecStart command worked!  I'm happy to announce that my Gemini server is now up and running.  Thanks so much!
